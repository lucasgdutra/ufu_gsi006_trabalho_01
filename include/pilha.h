/* --------------------------------------------------------------------------------------
   TAD Pilha
   Objetivo: Trabalho de disciplina Estrutura de dados (GSI006) BSI - Facom - UFU
   Restricoes:

   Autor: Dalmo Scalon Inacio/ Lucas Gabriel Dutra de Souza
   Disciplina: Estrutura de Dados 1
   Data da ultima atualizacao: 04/01/2023
----------------------------------------------------------------------------------------*/
#ifndef PILHA_H
#define PILHA_H

typedef struct node_pilha
{
    void *dados;
    struct node_pilha *proximo_node;
} Node_Pilha;

typedef struct pilha
{
    Node_Pilha *topo;
    int tamanho_pilha;
    int tamanho_estrutura_dados;
} Pilha;

/*inicializa estrutura pilha com valores iniciais, retorna 0 se não houver erro*/
int inicializa_pilha(Pilha *p, int tamanho_estrutura_dados);

/*adiciona elemento ao topo da pilha, retorna 0 se não houver erro*/
int empilha(Pilha *p, void *dados);

/*remove elemento do topo pilha e retorna valor removido pra ponteiro valor do argumento, retorna 0 se não houver erro*/
int desempilha(Pilha *p, void *valor);

/*remove todos elementos da pilha, retorna 0 se não houver erro*/
int limpa_pilha(Pilha *p);

/*copia tamanho da pilha para variavel tam em argumento, retorna 0 se não houver erro*/
int retorna_tamanho_pilha(Pilha *p, int *tam);

#endif