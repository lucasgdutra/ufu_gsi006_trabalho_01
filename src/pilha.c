#include "../include/pilha.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int inicializa_pilha(Pilha *p, int tamanho_estrutura_dados)
{
    p->topo = NULL;
    p->tamanho_estrutura_dados = tamanho_estrutura_dados;
    p->tamanho_pilha = 0;
    return 0;
}

int empilha(Pilha *p, void *dados)
{
    // instancia estrutura para receber elemento
    Node_Pilha *novo_node;
    novo_node = (Node_Pilha *)malloc(sizeof(Node_Pilha));
    if (novo_node == NULL)
    {
        // erro ao alocar elemento
        return 1;
    }
    novo_node->dados = malloc(p->tamanho_estrutura_dados);
    if (novo_node->dados == NULL)
    {
        // erro ao alocar dados do elemento
        free(novo_node);
        return 1;
    }
    novo_node->proximo_node = p->topo;
    // copia dados do argumento para novo node
    memcpy(novo_node->dados, dados, p->tamanho_estrutura_dados);

    // insere node no topo da pilha
    p->topo = novo_node;

    // ajusta tamanho da pilha
    p->tamanho_pilha++;
    return 0;
};

int desempilha(Pilha *p, void *valor)
{
    if (p->tamanho_pilha > 0)
    {
        // guarda endereço do node que vai ser removido
        Node_Pilha *node_removido = p->topo;

        // copia dados do node removido para ponteiro valor em argumento da função, se valor for NULL continua funcao sem copiar dados
        if (valor != NULL)
        {
            memcpy(valor, node_removido->dados, p->tamanho_estrutura_dados);
        }

        // redefine topo da pilha
        //
        p->topo = p->topo->proximo_node;


        // ajusta tamanho da pilha
        p->tamanho_pilha--;

        // libera memoria de node_removido
        free(node_removido->dados);
        free(node_removido);

        return 0;
    }
    else
    {
        return 1;
    }
}

int limpa_pilha(Pilha *p)
{
    int status = 0;
    while (p->tamanho_pilha > 0)
    {
        status = desempilha(p, NULL);
        if (status == 1)
        {
            return 1;
        }
    }
    return 0;
}

int retorna_tamanho_pilha(Pilha *p, int *tam)
{
    *tam = p->tamanho_pilha;
    return 0;
}