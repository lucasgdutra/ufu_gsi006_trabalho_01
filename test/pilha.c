#include "../include/pilha.h"
#include <stdlib.h>
#include <stdio.h>

int main(){
    Pilha p;
    inicializa_pilha(&p, sizeof(int));

    int entrada;
    while (scanf("%d", &entrada) != EOF) {
        int s = empilha(&p,&entrada);
        if (s == 1) return 1;
    }

    while (p.tamanho_pilha > 0) {
        int v;
        desempilha(&p,&v);
        printf("valor: %d\n",v);
    }

    return 0;
}