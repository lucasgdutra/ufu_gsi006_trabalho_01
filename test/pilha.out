valor: 11
valor: 9
valor: 7
valor: 5
valor: 3
valor: 1
